lib/utils/log/file_writer.go

```diff
@@ -0,0 +1,170 @@
+/*
+ * Teleport
+ * Copyright (C) 2024  Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package log
+
+import (
+	"context"
+	"io/fs"
+	"log/slog"
+	"os"
+	"path/filepath"
+	"sync"
+	"sync/atomic"
+
+	"github.com/fsnotify/fsnotify"
+	"github.com/gravitational/trace"
+)
+
+var (
+	// fileSharedWriter global file shared writer with the ability to reopen the log file
+	// in the same mode and with the same flags, either by function call or by filesystem
+	// notification subscription on the changes.
+	fileSharedWriter *FileSharedWriter
+
+	// mu protects global setter/getter of the file shared writer.
+	mu sync.Mutex
+)
+
+// FileSharedWriter is similar to SharedWriter except that it requires a `os.File` instead of a `io.Writer`.
+// This is to allow the File reopen required by logrotate and similar tools.
+type FileSharedWriter struct {
+	logFile  atomic.Pointer[os.File]
+	fileFlag int
+	fileMode fs.FileMode
+	watcher  *fsnotify.Watcher
+
+	lock sync.Mutex
+}
+
+// SetFileSharedWriter sets global file shared writer.
+func SetFileSharedWriter(sharedWriter *FileSharedWriter) (err error) {
+	mu.Lock()
+	defer mu.Unlock()
+
+	if fileSharedWriter != nil {
+		err = fileSharedWriter.Close()
+	}
+	fileSharedWriter = sharedWriter
+
+	return trace.Wrap(err)
+}
+
+// GetFileSharedWriter returns instance of the global file shared writer.
+func GetFileSharedWriter() *FileSharedWriter {
+	mu.Lock()
+	defer mu.Unlock()
+
+	return fileSharedWriter
+}
+
+// NewFileSharedWriter wraps the provided [os.File] in a writer that is thread safe,
+// with ability to enable filesystem notification watch and reopen file on specific events.
+func NewFileSharedWriter(logFile *os.File, flag int, mode fs.FileMode) (*FileSharedWriter, error) {
+	if logFile == nil {
+		return nil, trace.BadParameter("log file is not set")
+	}
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	sharedWriter := &FileSharedWriter{fileFlag: flag, fileMode: mode, watcher: watcher}
+	sharedWriter.logFile.Store(logFile)
+
+	return sharedWriter, nil
+}
+
+// Write writes len(b) bytes from b to the File.
+func (s *FileSharedWriter) Write(b []byte) (int, error) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	return s.logFile.Load().Write(b)
```

Comment:
If we acquire the lock in Write (which I guess we have to do anyway, just for the sake of serializing long messages from different logging operations) what's the point of keeping logFile in an atomic pointer? The only place where we interact with logFile without holding the lock is in the watcher loop, and it's only to read the file name, which we could just store in the FileSharedWriter instead, as it's never supposed to change (and, in fact, could probably just be sourced from the file name in the configuration to begin with).

Reply:
agree, I just left it more like preventive for future change, we have concurrent write in `reopen` and `read` in the watcher goroutine, and name never changing for now